### PR TITLE
Refactor strings column factories

### PR DESCRIPTION
**Pull request by harrism**
This PR refactors strings column factories to eliminate the use of `device_vector` and `std::vector` parameters, and to facility more use of `device_uvector` in calls to the factories. This is a small part of #7287 . Multiple versions of `make_strings_columns` take `device_vector` parameters. This PR expands the use of iterator and `device_span` versions to enable switching to `device_uvector` as described in #7287.  It also adds new `make_device_uvector_async/sync` utility functions.

This will help facilitate safe CUDA stream usage.

### CODE REVIEW

File: cpp/include/cudf/detail/utilities/vector_factories.hpp [hunk 1]
```diff
@@ -0,0 +1,242 @@
...
+ * container
+ *
+ * @note This function synchronizes `stream`.
+ *
+ * @tparam Container The type of the container to copy from
+ * @tparam T The type of the data to copy
+ * @param c The input device container from which to copy
+ * @param stream The stream on which to allocate memory and perform the copy
+ * @param mr The memory resource to use for allocating the returned device_uvector
+ * @return rmm::device_uvector<T> A device_uvector containing the copied data
```
Comment (davidwendt) [inline cpp/include/cudf/detail/utilities/vector_factories.hpp:226]:
```suggestion
 * @return  A device_uvector containing the copied data
```
Don't include the return type in `@return` doxygen tag.

File: cpp/include/cudf/detail/utilities/vector_factories.hpp [hunk 2]
```diff
@@ -0,0 +1,242 @@
...
+  Container<T> const& c,
+  rmm::cuda_stream_view stream        = rmm::cuda_stream_default,
+  rmm::mr::device_memory_resource* mr = rmm::mr::get_current_device_resource())
+{
+  return make_device_uvector_async(host_span<T const>{c}, stream, mr);
+}
+
+/**
+ * @brief Asynchronously construct a `device_uvector` containing a deep copy of data from a
+ * device_span
```
Comment (davidwendt) [inline cpp/include/cudf/detail/utilities/vector_factories.hpp:85]:
```suggestion
 * `device_span`
```

File: cpp/include/cudf/strings/detail/strings_column_factories.cuh [hunk 3]
```diff
@@ -25,6 +25,7 @@
 
 #include <rmm/cuda_stream_view.hpp>
 #include <rmm/exec_policy.hpp>
+#include "rmm/thrust_rmm_allocator.h"
```
Comment (davidwendt) [inline cpp/include/cudf/strings/detail/strings_column_factories.cuh:28]:
```suggestion
#include <rmm/thrust_rmm_allocator.h>
```

File: cpp/src/strings/strings_column_factories.cu [hunk 4]
```diff
@@ -18,92 +18,23 @@
...
 #include <strings/utilities.cuh>
 
 #include <rmm/cuda_stream_view.hpp>
 #include <rmm/device_vector.hpp>
 #include <rmm/exec_policy.hpp>
 
 #include <thrust/for_each.h>
+#include <thrust/iterator/counting_iterator.h>
+#include <thrust/iterator/zip_iterator.h>
 #include <thrust/transform_reduce.h>
```
Comment (davidwendt) [inline cpp/src/strings/strings_column_factories.cu:33]:
```suggestion
```
These headers are not used by this `.cu` file.

File: cpp/src/strings/strings_column_factories.cu [hunk 5]
```diff
@@ -18,92 +18,23 @@
 #include <cudf/column/column_factories.hpp>
 #include <cudf/detail/nvtx/ranges.hpp>
 #include <cudf/detail/valid_if.cuh>
```
Comment (davidwendt) [inline cpp/src/strings/strings_column_factories.cu:20]:
```suggestion
```
This header is not needed anymore.

File: cpp/src/strings/strings_column_factories.cu [hunk 6]
```diff
@@ -18,92 +18,23 @@
 #include <cudf/column/column_factories.hpp>
 #include <cudf/detail/nvtx/ranges.hpp>
```
Comment (davidwendt) [inline cpp/src/strings/strings_column_factories.cu:19]: You removed all of the `CUDF_FUNC_RANGE();` calls to the `detail::make_strings_column`. Usually these are put in the non-detail function right before calling the detail version.

**Review by davidwendt — CHANGES_REQUESTED** [review]

File: cpp/include/cudf/detail/utilities/vector_factories.hpp (hunk 1 above)
Comment (harrism) [reply cpp/include/cudf/detail/utilities/vector_factories.hpp:226]: Ugh, my vscode plugin does that automatically. :(  Will remove.

**Review by ajschmidt8 — APPROVED** [review]

**Comment by harrism** [comment]
@davidwendt I addressed your feedback and also added some more device_vector -> uvector changes in strings (the simple ones I found). I think that's enough for this PR though.

File: cpp/src/strings/findall.cu [hunk 7]
```diff
@@ -164,29 +164,28 @@ std::unique_ptr<table> findall_re(
       strings_count));
 
   for (int32_t column_index = 0; column_index < columns; ++column_index) {
-    rmm::device_vector<string_index_pair> indices(strings_count);
-    string_index_pair* d_indices = indices.data().get();
+    rmm::device_uvector<string_index_pair> indices(strings_count, stream);
```
Comment (davidwendt) [inline cpp/src/strings/findall.cu:167]: I'm making some more involved changes in this file in #7428

File: cpp/src/strings/split/split.cu [hunk 8]
```diff
@@ -506,8 +506,8 @@ std::unique_ptr<table> split_fn(strings_column_view const& strings_column,
   }
 
   // create working area to hold all token positions
-  rmm::device_vector<string_index_pair> tokens(columns_count * strings_count);
-  string_index_pair* d_tokens = tokens.data().get();
+  rmm::device_uvector<string_index_pair> tokens(columns_count * strings_count, stream);
+  string_index_pair* d_tokens = tokens.data();
```
Comment (davidwendt) [inline cpp/src/strings/split/split.cu:510]: I'm making more involved changes to this file in #7427

**Review by davidwendt — CHANGES_REQUESTED** [review]
Some of the changes in this PR are more than necessary for refactoring the strings column factory. Two of the non-refactory changes become a conflict for at least two of my PRs. Request maybe this PR can be broken up.

File: cpp/include/cudf/column/column_factories.hpp [hunk 9]
```diff
@@ -344,20 +345,19 @@ std::unique_ptr<column> make_fixed_width_column(
...
- * (indicating a null string). The size must be the number of bytes.
+ * @param[in] strings The device span of pointer/size pairs. Each pointer must be a device memory
+   address or `nullptr` (indicating a null string). The size must be the number of bytes.
  * @param[in] stream CUDA stream used for device memory operations and kernel launches.
  * @param[in] mr Device memory resource used for allocation of the column's `null_mask` and children
  * columns' device memory.
  */
 std::unique_ptr<column> make_strings_column(
-  const rmm::device_vector<thrust::pair<const char*, size_type>>& strings,
+  cudf::device_span<thrust::pair<const char*, size_type>> strings,
```
Comment (jrhemstad) [inline cpp/include/cudf/column/column_factories.hpp:355]: I think you want a `const` on the pair type too. 
```suggestion
  cudf::device_span<thrust::pair<const char*, size_type> const> strings,
```

File: cpp/include/cudf/strings/detail/strings_column_factories.cuh [hunk 10]
```diff
@@ -95,6 +95,50 @@ std::unique_ptr<column> make_strings_column(IndexPairIterator begin,
...
+                                            rmm::device_buffer&& null_mask,
+                                            rmm::cuda_stream_view stream,
+                                            rmm::mr::device_memory_resource* mr)
+{
+  CUDF_FUNC_RANGE();
+  size_type strings_count = thrust::distance(offsets_begin, offsets_end) - 1;
+  size_type bytes         = std::distance(chars_begin, chars_end) * sizeof(char);
+  if (strings_count == 0) return strings::detail::make_empty_strings_column(stream, mr);
+
+  CUDF_EXPECTS(null_count < strings_count, "null strings column not yet supported");
```
Comment (jrhemstad) [inline cpp/include/cudf/strings/detail/strings_column_factories.cuh:114]: Really?

File: cpp/src/io/json/reader_impl.cu [hunk 11]
```diff
@@ -600,9 +601,22 @@ table_with_metadata reader::impl::convert_data_to_table(rmm::cuda_stream_view st
...
+  auto repl_chars   = std::vector<char>{'"', '\\', '\t', '\r', '\b'};
+  auto repl_offsets = std::vector<size_type>{0, 1, 2, 3, 4, 5};
+
+  auto target = make_strings_column(cudf::detail::make_device_uvector_async(target_chars, stream),
+                                    cudf::detail::make_device_uvector_async(target_offsets, stream),
+                                    {},
+                                    0,
+                                    stream);
+  auto repl   = make_strings_column(cudf::detail::make_device_uvector_async(repl_chars, stream),
+                                  cudf::detail::make_device_uvector_async(repl_offsets, stream),
```
Comment (jrhemstad) [inline cpp/src/io/json/reader_impl.cu:616]: Is there a sync somewhere later to make sure this copy is done?

File: cpp/src/io/json/reader_impl.cu [hunk 12]
```diff
@@ -600,9 +601,22 @@ table_with_metadata reader::impl::convert_data_to_table(rmm::cuda_stream_view st
...
+                                  {},
+                                  0,
+                                  stream);
 
   thrust::host_vector<cudf::size_type> h_valid_counts = d_valid_counts;
```
Comment (jrhemstad) [inline cpp/src/io/json/reader_impl.cu:621]: This is doing a D2H copy w/o a stream.

File: cpp/include/cudf/detail/utilities/vector_factories.hpp [hunk 13]
```diff
@@ -0,0 +1,242 @@
...
+  rmm::mr::device_memory_resource* mr = rmm::mr::get_current_device_resource())
+{
+  rmm::device_uvector<T> ret(source_data.size(), stream, mr);
+  CUDA_TRY(cudaMemcpyAsync(ret.data(),
+                           source_data.data(),
+                           source_data.size() * sizeof(T),
+                           cudaMemcpyDefault,
+                           stream.value()));
+  stream.synchronize();
+  return ret;
```
Comment (jrhemstad) [inline cpp/include/cudf/detail/utilities/vector_factories.hpp:160]: This can just call the `async` version and then call `sync`, right?

File: cpp/include/cudf/detail/utilities/vector_factories.hpp [hunk 14]
```diff
@@ -0,0 +1,242 @@
...
```
Comment (jrhemstad) [inline cpp/include/cudf/detail/utilities/vector_factories.hpp:212]: Same here. Use the async version to implement the sync version.

File: cpp/include/cudf/detail/utilities/vector_factories.hpp [hunk 15]
```diff
@@ -0,0 +1,242 @@
...
+template <template <typename, typename...> typename Container,
+          typename T,
+          std::enable_if_t<std::is_convertible<Container<T>, host_span<T const>>::value>* = nullptr>
+rmm::device_uvector<T> make_device_uvector_async(
+  Container<T> const& c,
+  rmm::cuda_stream_view stream        = rmm::cuda_stream_default,
+  rmm::mr::device_memory_resource* mr = rmm::mr::get_current_device_resource())
+{
+  return make_device_uvector_async(host_span<T const>{c}, stream, mr);
+}
```
Comment (jrhemstad) [inline cpp/include/cudf/detail/utilities/vector_factories.hpp:81]: This is a slightly cleaner way to do it. 
```suggestion
template <typename Container,
          std::enable_if_t<std::is_convertible<Container, host_span<Container::value_type const>>::value>* = nullptr>
rmm::device_uvector<Container::value_type> make_device_uvector_async(
  Container const& c,
  rmm::cuda_stream_view stream        = rmm::cuda_stream_default,
  rmm::mr::device_memory_resource* mr = rmm::mr::get_current_device_resource())
{
  return make_device_uvector_async(host_span<Container::value_type const>{c}, stream, mr);
}
```

**Comment by harrism** [comment]
> Some of the changes in this PR are more than necessary for refactoring the strings column factory. Two of the non-refactory changes become a conflict for at least two of my PRs. Request maybe this PR can be broken up.

The goal of the PR is ultimately to replave device_vector... If you just merge your PRs first I will resolve conflicts.

File: cpp/src/io/json/reader_impl.cu (hunk 12 above)
Comment (harrism) [reply cpp/src/io/json/reader_impl.cu:621]: There are many other bugs that this PR does not fix as well! :)

File: cpp/include/cudf/detail/utilities/vector_factories.hpp (hunk 13 above)
Comment (harrism) [reply cpp/include/cudf/detail/utilities/vector_factories.hpp:160]: Yeah good point.

File: cpp/src/strings/findall.cu (hunk 7 above)
Comment (harrism) [reply cpp/src/strings/findall.cu:168]: Resolved.

File: cpp/src/strings/split/split.cu (hunk 8 above)
Comment (harrism) [reply cpp/src/strings/split/split.cu]: Resolved (there weren't conflicts in this file).

File: cpp/src/io/json/reader_impl.cu (hunk 11 above)
Comment (harrism) [reply cpp/src/io/json/reader_impl.cu:616]: Ignoring the later D2H copy without a stream, I don't think this requires a sync, since the function we are inside takes a stream, and is thus assumed to follow stream ordering semantics on that stream (and does so by passing the stream to these uvector and make_strings_column functions.

File: cpp/src/io/json/reader_impl.cu (hunk 12 above)
Comment (harrism) [reply cpp/src/io/json/reader_impl.cu:621]: This function also uses a log of device_vectors. Replacing them is non-trivial (unlike some of the strings changes I made), so I prefer to tackle that in a separate PR.

File: cpp/include/cudf/column/column_factories.hpp (hunk 9 above)
Comment (harrism) [reply cpp/include/cudf/column/column_factories.hpp]: Added, thanks.

File: cpp/include/cudf/detail/utilities/vector_factories.hpp (hunk 13 above)
Comment (harrism) [reply cpp/include/cudf/detail/utilities/vector_factories.hpp]: Done.

File: cpp/include/cudf/detail/utilities/vector_factories.hpp (hunk 14 above)
Comment (harrism) [reply cpp/include/cudf/detail/utilities/vector_factories.hpp]: Done.

File: cpp/include/cudf/detail/utilities/vector_factories.hpp (hunk 15 above)
Comment (harrism) [reply cpp/include/cudf/detail/utilities/vector_factories.hpp]: Marginally, since you have to use the `typename` keyword everywhere you have `Container::value_type` it's just about the same verbosity.  But I made the change.

File: cpp/src/io/json/reader_impl.cu (hunk 11 above)
Comment (jrhemstad) [reply cpp/src/io/json/reader_impl.cu:616]: But the `target_chars/target_offsets` and `repl_chars/repl_offsets` are temporary local vectors that will get destroyed when the function exits and the copy may not have completed yet. Or am I missing something?

**Review by cwharris — CHANGES_REQUESTED** [review]

File: cpp/include/cudf/column/column_factories.hpp [hunk 16]
```diff
@@ -330,7 +331,7 @@ std::unique_ptr<column> make_fixed_width_column(
 }
 
 /**
- * @brief Construct STRING type column given a device vector of pointer/size pairs.
+ * @brief Construct STRING type column given a device span of pointer/size pairs.
```
Comment (davidwendt) [inline cpp/include/cudf/column/column_factories.hpp:334]: Add a blank line after the `@brief` tag.
```suggestion
 * @brief Construct STRING type column given a device span of pointer/size pairs.
 
```
https://github.com/rapidsai/cudf/blob/branch-0.19/cpp/docs/DOCUMENTATION.md#brief

File: cpp/include/cudf/column/column_factories.hpp [hunk 17]
```diff
@@ -344,20 +345,19 @@ std::unique_ptr<column> make_fixed_width_column(
...
  */
 std::unique_ptr<column> make_strings_column(
-  const rmm::device_vector<thrust::pair<const char*, size_type>>& strings,
+  cudf::device_span<thrust::pair<const char*, size_type> const> strings,
   rmm::cuda_stream_view stream        = rmm::cuda_stream_default,
   rmm::mr::device_memory_resource* mr = rmm::mr::get_current_device_resource());
 
 /**
- * @brief Construct STRING type column given a device vector of string_view.
+ * @brief Construct STRING type column given a device span of string_view.
```
Comment (davidwendt) [inline cpp/include/cudf/column/column_factories.hpp:360]:
```suggestion
 * @brief Construct STRING type column given a device span of string_view.
 
```

File: cpp/include/cudf/column/column_factories.hpp [hunk 18]
```diff
@@ -372,118 +372,68 @@ std::unique_ptr<column> make_strings_column(
...
+  size_type null_count                            = cudf::UNKNOWN_NULL_COUNT,
+  rmm::cuda_stream_view stream                    = rmm::cuda_stream_default,
+  rmm::mr::device_memory_resource* mr             = rmm::mr::get_current_device_resource());
 
 /**
- * @brief Constructs a STRING type column given offsets column, chars columns,
- * and null mask and null count. The columns and mask are moved into the
- * resulting strings column.
+ * @brief Constructs a STRING type column given offsets column, chars columns, and null mask and
+ *  null count. The columns and mask are moved into the resulting strings column.
```
Comment (davidwendt) [inline cpp/include/cudf/column/column_factories.hpp:426]:
```suggestion
 *  null count. 
 *
 * The columns and mask are moved into the resulting strings column.
```

File: cpp/tests/strings/combine_tests.cpp [hunk 19]
```diff
@@ -281,8 +281,13 @@ TEST_F(StringsConcatenateWithColSeparatorTest, SingleColumnStringMixNoReplacemen
     {"", "", "", "bbabc", "", "d", "éa", "", "bbb", "éééf"},
     {false, false, true, true, false, true, true, false, true, true});
 
+  CUDA_TRY(cudaDeviceSynchronize());
```
Comment (davidwendt) [inline cpp/tests/strings/combine_tests.cpp:284]: Why is this needed here?

File: cpp/tests/strings/combine_tests.cpp [hunk 20]
```diff
@@ -281,8 +281,13 @@ TEST_F(StringsConcatenateWithColSeparatorTest, SingleColumnStringMixNoReplacemen
...
+
   auto results =
     cudf::strings::concatenate(cudf::table_view{{col0}}, cudf::strings_column_view(sep_col));
+
+  CUDA_TRY(cudaDeviceSynchronize());
```
Comment (davidwendt) [inline cpp/tests/strings/combine_tests.cpp:289]: I don't think this is needed either.

**Review by davidwendt — CHANGES_REQUESTED** [review]

File: cpp/src/io/json/reader_impl.cu (hunk 11 above)
Comment (harrism) [reply cpp/src/io/json/reader_impl.cu:616]: Changing the structure is out of scope of this PR, so I added a sync for now with a comment explaining why.

File: cpp/include/cudf/column/column_factories.hpp (hunk 16 above)
Comment (harrism) [reply cpp/include/cudf/column/column_factories.hpp:334]: I think the blank line needs to start with a `*`.

File: cpp/tests/strings/combine_tests.cpp (hunk 19 above)
Comment (harrism) [reply cpp/tests/strings/combine_tests.cpp:284]: Good catch, must be leftover from debugging

**Comment by harrism** [comment]
@cwharris what changes are you requesting?

**Review by jrhemstad — APPROVED** [review]

File: cpp/tests/strings/combine_tests.cpp [hunk 21]
```diff
@@ -283,6 +283,7 @@ TEST_F(StringsConcatenateWithColSeparatorTest, SingleColumnStringMixNoReplacemen
 
   auto results =
     cudf::strings::concatenate(cudf::table_view{{col0}}, cudf::strings_column_view(sep_col));
+
```
Comment (davidwendt) [inline cpp/tests/strings/combine_tests.cpp:286]: Maybe revert this file. You are only adding a blank line now.

File: cpp/include/cudf/detail/utilities/vector_factories.hpp [hunk 22]
```diff
@@ -0,0 +1,237 @@
+/*
+ *
```
Comment (davidwendt) [inline cpp/include/cudf/detail/utilities/vector_factories.hpp:2]:
```suggestion

```

**Review by davidwendt — CHANGES_REQUESTED** [review]
Some of the files need 2021 added to their copyright line.

File: cpp/include/cudf/column/column_factories.hpp (hunk 16 above)
Comment (davidwendt) [reply cpp/include/cudf/column/column_factories.hpp]: Oops, right.

**Comment by harrism** [comment]
> Some of the files need 2021 added to their copyright line.

Thanks. Done.

**Review by davidwendt — APPROVED** [review]

**Comment by shwina** [comment]
rerun tests

**Review by cwharris — APPROVED** [review]

**Comment by cwharris** [comment]
@harrism you addressed my concerns. :+1:

**Comment by harrism** [comment]
@gpucibot merge
